File: lib/src/dnd/dnd.dart
import '../encoding/encoding_util.dart';
import '../encoding/token.dart';

/// Drag & drop interaction state. Mirrors `SC.DROP_STATE`.
class DropState {
  static const idle = 'idle';
  static const dragOver = 'dragOver';
  static const rejected = 'rejected';
  static const dropped = 'dropped';
}

/// A normalized dropped resource. Mirrors `SmartComposerDropItem`.
class DropItem {
  DropItem({
    required this.type,
    required this.name,
    this.path = '',
    required this.uri,
    this.mimeType = '',
    this.extension = '',
    this.size = 0,
    this.bytes,
    Map<String, dynamic>? metadata,
    this.source = 'os-file',
    this.isLocal = true,
    this.isRemote = false,
    this.note,
  }) : metadata = metadata ?? <String, dynamic>{};

  final String type;
  final String name;
  final String path;
  final String uri;
  final String mimeType;
  final String extension;
  final int size;
  final List<int>? bytes;
  final Map<String, dynamic> metadata;
  final String source;
  final bool isLocal;
  final bool isRemote;

  /// Optional UI note used by demo tray chips (e.g. "over 25 MB → rejected").
  final String? note;
}

/// Configuration for drag & drop. Mirrors `SmartComposerDropConfig` /
/// `SC.DEFAULT_DROP_CONFIG`.
class DropConfig {
  const DropConfig({
    this.enabled = true,
    this.allowMultiple = true,
    this.allowedExtensions,
    this.allowedMimeTypes,
    this.blockedExtensions = const ['exe', 'bat', 'cmd', 'sh', 'dll', 'msi', 'app'],
    this.maxFileSize = 25 * 1024 * 1024,
    this.maxFilesCount = 10,
    this.insertAtDropPosition = true,
    this.fallbackInsertAtEnd = true,
    this.replaceSelectionOnDrop = true,
    this.generateTokenFromDropItem,
    this.customValidator,
  });

  final bool enabled;
  final bool allowMultiple;
  final List<String>? allowedExtensions;
  final List<String>? allowedMimeTypes;
  final List<String> blockedExtensions;
  final int maxFileSize;
  final int maxFilesCount;
  final bool insertAtDropPosition;
  final bool fallbackInsertAtEnd;
  final bool replaceSelectionOnDrop;
  final SmartComposerToken Function(DropItem item)? generateTokenFromDropItem;
  final DropValidationResult Function(DropItem item)? customValidator;

  DropConfig copyWith({
    bool? allowMultiple,
    List<String>? allowedExtensions,
    int? maxFileSize,
  }) {
    return DropConfig(
      enabled: enabled,
      allowMultiple: allowMultiple ?? this.allowMultiple,
      allowedExtensions: allowedExtensions ?? this.allowedExtensions,
      allowedMimeTypes: allowedMimeTypes,
      blockedExtensions: blockedExtensions,
      maxFileSize: maxFileSize ?? this.maxFileSize,
      maxFilesCount: maxFilesCount,
      insertAtDropPosition: insertAtDropPosition,
      fallbackInsertAtEnd: fallbackInsertAtEnd,
      replaceSelectionOnDrop: replaceSelectionOnDrop,
      generateTokenFromDropItem: generateTokenFromDropItem,
      customValidator: customValidator,
    );
  }
}

/// One validation error for a dropped item.
class DropError {
  const DropError({required this.code, required this.message});
  final String code;
  final String message;
}

/// Result of validating one dropped item. Mirrors `{ valid, errors, item }`.
class DropValidationResult {
  const DropValidationResult({required this.valid, required this.errors, this.item});
  final bool valid;
  final List<DropError> errors;
  final DropItem? item;
}

/// Turns dropped files / URLs / custom resources into typed smart tokens.
/// Faithful port of `dnd.js`.
class ComposerDnd {
  ComposerDnd._();

  static const _ext = {
    'image': ['png', 'jpg', 'jpeg', 'gif', 'webp', 'svg', 'bmp', 'heic', 'avif'],
    'video': ['mp4', 'mov', 'avi', 'mkv', 'webm', 'm4v'],
    'document': ['pdf', 'doc', 'docx', 'xls', 'xlsx', 'ppt', 'pptx', 'txt', 'csv', 'md', 'rtf', 'pages', 'numbers', 'key'],
  };

  /// Extension / MIME → tagType. Mirrors `SC.mapExtToType`.
  static String mapExtToType(String? ext, [String? mime]) {
    final e = (ext ?? '').toLowerCase();
    final m = (mime ?? '').toLowerCase();
    if (m.startsWith('image/') || _ext['image']!.contains(e)) return 'image';
    if (m.startsWith('video/') || _ext['video']!.contains(e)) return 'video';
    if (_ext['document']!.contains(e) ||
        RegExp(r'pdf|word|excel|powerpoint|spreadsheet|presentation|msword|officedocument|csv|text\/plain').hasMatch(m)) {
      return 'document';
    }
    return 'file';
  }

  static String _extOf(String name) {
    final m = RegExp(r'\.([A-Za-z0-9]+)$').firstMatch(name);
    return m != null ? m.group(1)!.toLowerCase() : '';
  }

  static String _basename(String p) => p.split(RegExp(r'[\\/]')).last;

  /// Build a safe file URI from a path or bare name. Mirrors the private
  /// `fileUri()` in `dnd.js`.
  static String fileUri(String pathOrName) {
    final p = pathOrName;
    if (RegExp(r'^[a-z][\w+.-]*:\/\/', caseSensitive: false).hasMatch(p)) return p;
    if (RegExp(r'^[a-zA-Z]:[\\/]').hasMatch(p)) return 'file:///${p.replaceAll('\\', '/')}';
    if (p.startsWith('/')) return 'file://$p';
    return 'file:///${Uri.encodeFull(p).replaceAll(RegExp(r'^/+'), '')}';
  }

  /// Build a [DropItem] from a raw map. Mirrors `SC.makeDropItem`.
  static DropItem makeDropItem(Map<String, dynamic> raw) {
    final name = (raw['name'] as String?) ??
        (raw['path'] != null ? _basename(raw['path'] as String) : null) ??
        (raw['uri'] != null ? _basename(raw['uri'] as String) : null) ??
        'file';
    final ext = (raw['extension'] as String?) ?? _extOf(name);
    final type = (raw['type'] as String?) ?? mapExtToType(ext, raw['mimeType'] as String?);
    return DropItem(
      type: type,
      name: name,
      path: (raw['path'] as String?) ?? '',
      uri: (raw['uri'] as String?) ?? fileUri((raw['path'] as String?) ?? name),
      mimeType: (raw['mimeType'] as String?) ?? '',
      extension: ext,
      size: (raw['size'] as num?)?.round() ?? 0,
      bytes: (raw['bytes'] as List?)?.cast<int>(),
      metadata: (raw['metadata'] as Map?)?.cast<String, dynamic>(),
      source: (raw['source'] as String?) ?? 'os-file',
      isLocal: raw['isLocal'] != false,
      isRemote: raw['isRemote'] == true,
      note: raw['_note'] as String?,
    );
  }

  /// Build a [DropItem] from a URL. Mirrors `SC.makeDropItemFromUrl`.
  static DropItem makeDropItemFromUrl(String url) {
    final clean = url.trim();
    final label = clean.replaceAll(RegExp(r'^https?:\/\/'), '').replaceAll(RegExp(r'\/$'), '');
    return makeDropItem({
      'type': 'link',
      'name': label,
      'uri': clean,
      'mimeType': 'text/uri-list',
      'source': 'url',
      'isLocal': false,
      'isRemote': true,
    });
  }

  /// Dropped item → encoded token. Mirrors `SC.dropItemToToken`.
  static SmartComposerToken dropItemToToken(DropItem item) {
    return SmartComposerToken(
      prefix: EncodingUtil.prefixForType(item.type),
      tagType: item.type,
      displayText: item.name,
      valueText: item.uri,
      rawText: '',
      metadata: item.metadata,
      resolveState: ResolveState.idle,
    );
  }

  /// Human-readable byte size. Mirrors `SC.formatBytes`.
  static String formatBytes(num n) {
    if (n < 1024) return '${n.round()} B';
    if (n < 1024 * 1024) return '${(n / 1024).round()} KB';
    return '${(n / 1024 / 1024).toStringAsFixed(1)} MB';
  }
}

/// Validates a dropped item against a [DropConfig]. Mirrors
/// `SC.SmartComposerDropValidator`.
class SmartComposerDropValidator {
  SmartComposerDropValidator._();

  static bool _mimeOk(String mime, List<String> list) {
    return list.any((p) => p == mime || (p.endsWith('/*') && mime.startsWith(p.substring(0, p.length - 1))));
  }

  static DropValidationResult validate(DropItem item, DropConfig cfg) {
    final errors = <DropError>[];
    final ext = item.extension.toLowerCase();
    if (cfg.blockedExtensions.contains(ext)) {
      errors.add(DropError(code: 'blockedType', message: '.$ext files are not allowed.'));
    }
    if (cfg.allowedExtensions != null && ext.isNotEmpty && !cfg.allowedExtensions!.contains(ext)) {
      errors.add(DropError(code: 'extensionNotAllowed', message: '.$ext is not an accepted type.'));
    }
    if (cfg.allowedMimeTypes != null && item.mimeType.isNotEmpty && !_mimeOk(item.mimeType, cfg.allowedMimeTypes!)) {
      errors.add(DropError(code: 'mimeNotAllowed', message: '${item.mimeType} is not accepted.'));
    }
    if (item.size > 0 && item.size > cfg.maxFileSize) {
      errors.add(DropError(
        code: 'tooLarge',
        message: '${ComposerDnd.formatBytes(item.size)} exceeds the ${ComposerDnd.formatBytes(cfg.maxFileSize)} limit.',
      ));
    }
    if (cfg.customValidator != null) {
      final r = cfg.customValidator!(item);
      if (!r.valid) {
        errors.addAll(r.errors.isNotEmpty ? r.errors : const [DropError(code: 'custom', message: 'Rejected.')]);
      }
    }
    return DropValidationResult(valid: errors.isEmpty, errors: errors, item: item);
  }
}

File: docs/composer/dnd.js
/* =============================================================================
   SmartComposer — DRAG & DROP  (framework-agnostic; no React)
   -----------------------------------------------------------------------------
   Turns dropped files / images / videos / documents / URLs / custom resources
   into encoded smart tokens — [<prefix><tagType>:<displayText>](valueText).
   The editor never uploads anything and never hardcodes backend logic; it only
   builds tokens and emits callbacks. All shapes mirror the spec models.

   Exposes:
     SC.DEFAULT_DROP_CONFIG
     SC.makeDropItemFromFile(file) / makeDropItemFromUrl(url) / makeDropItem(raw)
     SC.mapExtToType(ext, mime)
     SC.dropItemToToken(item)            // SmartComposerDropItem -> SmartComposerToken
     SC.SmartComposerDropValidator.validate(item, config)
     SC.DROP_STATE
============================================================================= */
(function () {
  const SC = (window.SC = window.SC || {});

  SC.DROP_STATE = Object.freeze({ idle: 'idle', dragOver: 'dragOver', rejected: 'rejected', dropped: 'dropped' });

  /* ---- default config (SmartComposerDropConfig) ---- */
  SC.DEFAULT_DROP_CONFIG = {
    enabled: true,
    allowMultiple: true,
    allowedExtensions: null,     // e.g. ['pdf','png'] — null = any
    allowedMimeTypes: null,      // e.g. ['image/*'] — null = any
    blockedExtensions: ['exe', 'bat', 'cmd', 'sh', 'dll', 'msi', 'app'],
    maxFileSize: 25 * 1024 * 1024,  // 25 MB
    maxFilesCount: 10,
    insertAtDropPosition: true,
    fallbackInsertAtEnd: true,
    replaceSelectionOnDrop: true,
    // generateTokenFromDropItem / customValidator can be supplied by the app
  };

  /* ---- extension / mime → tagType ---- */
  const EXT = {
    image: ['png', 'jpg', 'jpeg', 'gif', 'webp', 'svg', 'bmp', 'heic', 'avif'],
    video: ['mp4', 'mov', 'avi', 'mkv', 'webm', 'm4v'],
    document: ['pdf', 'doc', 'docx', 'xls', 'xlsx', 'ppt', 'pptx', 'txt', 'csv', 'md', 'rtf', 'pages', 'numbers', 'key'],
  };
  SC.mapExtToType = function (ext, mime) {
    ext = (ext || '').toLowerCase();
    mime = (mime || '').toLowerCase();
    if (mime.startsWith('image/') || EXT.image.includes(ext)) return 'image';
    if (mime.startsWith('video/') || EXT.video.includes(ext)) return 'video';
    if (EXT.document.includes(ext) || /pdf|word|excel|powerpoint|spreadsheet|presentation|msword|officedocument|csv|text\/plain/.test(mime)) return 'document';
    return 'file';
  };

  function extOf(name) { const m = /\.([A-Za-z0-9]+)$/.exec(name || ''); return m ? m[1].toLowerCase() : ''; }
  function basename(p) { return String(p || '').split(/[\\/]/).pop(); }

  // Build a safe file URI. Browsers can't read absolute OS paths from a real
  // drop (security), so we fall back to a name-based file:/// URI; a richer
  // host (Electron / native) can pass a full path which we preserve.
  function fileUri(pathOrName) {
    const p = pathOrName || '';
    if (/^[a-z][\w+.-]*:\/\//i.test(p)) return p;             // already a URI
    if (/^[a-zA-Z]:[\\/]/.test(p)) return 'file:///' + p.replace(/\\/g, '/'); // windows path
    if (p.startsWith('/')) return 'file://' + p;               // posix absolute
    return 'file:///' + encodeURI(p).replace(/^\/+/, '');      // bare name
  }

  /* ---- SmartComposerDropItem factory ---- */
  SC.makeDropItem = function (raw) {
    const name = raw.name || basename(raw.path) || basename(raw.uri) || 'file';
    const ext = raw.extension || extOf(name);
    const type = raw.type || SC.mapExtToType(ext, raw.mimeType);
    return {
      type, name,
      path: raw.path || '',
      uri: raw.uri || fileUri(raw.path || name),
      mimeType: raw.mimeType || '',
      extension: ext,
      size: raw.size != null ? raw.size : 0,
      bytes: raw.bytes || null,
      metadata: raw.metadata || {},
      source: raw.source || 'os-file',
      isLocal: raw.isLocal !== false,
      isRemote: !!raw.isRemote,
    };
  };

  SC.makeDropItemFromFile = function (file) {
    const ext = extOf(file.name);
    // file.path exists only in Electron-like hosts; webkitRelativePath in dir picks
    const p = file.path || file.webkitRelativePath || file.name;
    return SC.makeDropItem({
      type: SC.mapExtToType(ext, file.type), name: file.name, path: file.path || '',
      uri: fileUri(p), mimeType: file.type || '', extension: ext, size: file.size,
      source: 'os-file', isLocal: true,
    });
  };

  SC.makeDropItemFromUrl = function (url) {
    const clean = String(url).trim();
    const label = clean.replace(/^https?:\/\//, '').replace(/\/$/, '');
    return SC.makeDropItem({
      type: 'link', name: label, uri: clean, mimeType: 'text/uri-list',
      source: 'url', isLocal: false, isRemote: true,
    });
  };

  /* ---- dropped item → encoded token ---- */
  SC.dropItemToToken = function (item) {
    return {
      prefix: SC.prefixForType(item.type),
      tagType: item.type,
      displayText: item.name,
      valueText: item.uri,
      rawText: '',
      metadata: item.metadata || {},
      resolveState: SC.RESOLVE_STATE ? SC.RESOLVE_STATE.idle : 'idle',
    };
  };

  /* ---- validator (SmartComposerDropValidator) ---- */
  SC.SmartComposerDropValidator = {
    validate(item, cfg) {
      cfg = Object.assign({}, SC.DEFAULT_DROP_CONFIG, cfg || {});
      const errors = [];
      const ext = (item.extension || '').toLowerCase();
      if (cfg.blockedExtensions && cfg.blockedExtensions.includes(ext))
        errors.push({ code: 'blockedType', message: `.${ext} files are not allowed.` });
      if (cfg.allowedExtensions && ext && !cfg.allowedExtensions.includes(ext))
        errors.push({ code: 'extensionNotAllowed', message: `.${ext} is not an accepted type.` });
      if (cfg.allowedMimeTypes && item.mimeType && !mimeOk(item.mimeType, cfg.allowedMimeTypes))
        errors.push({ code: 'mimeNotAllowed', message: `${item.mimeType} is not accepted.` });
      if (cfg.maxFileSize && item.size && item.size > cfg.maxFileSize)
        errors.push({ code: 'tooLarge', message: `${fmtSize(item.size)} exceeds the ${fmtSize(cfg.maxFileSize)} limit.` });
      if (cfg.customValidator) {
        const r = cfg.customValidator(item);
        if (r && r.valid === false) errors.push(...(r.errors || [{ code: 'custom', message: 'Rejected.' }]));
      }
      return { valid: errors.length === 0, errors, item };
    },
  };
  function mimeOk(mime, list) {
    return list.some((p) => p === mime || (p.endsWith('/*') && mime.startsWith(p.slice(0, -1))));
  }
  function fmtSize(n) {
    if (n < 1024) return n + ' B';
    if (n < 1024 * 1024) return (n / 1024).toFixed(0) + ' KB';
    return (n / 1024 / 1024).toFixed(1) + ' MB';
  }
  SC.formatBytes = fmtSize;

  /* ---- read a DataTransfer into SmartComposerDropItems ----
     Priority: app-custom payload > OS files > URL/text. Returns [] if nothing. */
  SC.dropItemsFromDataTransfer = function (dt) {
    const items = [];
    // 1) custom in-app drag payload (draggable chips, resource lists, etc.)
    try {
      const custom = dt.getData('application/x-smartcomposer');
      if (custom) {
        const arr = JSON.parse(custom);
        (Array.isArray(arr) ? arr : [arr]).forEach((r) => items.push(SC.makeDropItem(r)));
        if (items.length) return items;
      }
    } catch (e) { /* ignore malformed */ }
    // 2) real OS files
    if (dt.files && dt.files.length) {
      for (let i = 0; i < dt.files.length; i++) items.push(SC.makeDropItemFromFile(dt.files[i]));
      if (items.length) return items;
    }
    // 3) URL / text
    const uri = dt.getData('text/uri-list') || dt.getData('text/plain');
    if (uri && /^(https?|ftp|mailto):/i.test(uri.trim())) {
      uri.split(/\r?\n/).filter((l) => l && !l.startsWith('#')).forEach((u) => items.push(SC.makeDropItemFromUrl(u)));
    }
    return items;
  };
})();
